gpflow/optimizers/scipy.py

```diff
@@ -7,16 +7,15 @@
 
 __all__ = ["Scipy"]
 
-Variables = Iterable[tf.Variable]
-StepCallback = Callable[[int, Variables, List[tf.Tensor]], None]
+StepCallback = Callable[[int, List[tf.Variable], List[tf.Tensor]], None]
```

Comment:
you've replaced `Variables` with `List[tf.Variable]` in a number of places. This is an API change (though admittedly a small one). was `Variables` a bug?

Reply:
It depends which way around you see it: the step_callback will be passed lists - does the type signature need to reflect what the types of the arguments are going to be, or what types of arguments it can accept? If the latter, happy to change this to
```suggestion
StepCallback = Callable[[int, Iterable[tf.Variable], Iterable[tf.Tensor]], None]
```
it should be consistent for both second and third argument though!

Reply:
types _usually_ reflect what the types need to be for the functionality to work. There are exceptions, such as where there's a good change you'll need to constrain the types in the future and don't want to prematurely make compatibility promises that you'll need to retract later.

Reply:
in this case, if it's valid for it to be `Iterable`, it should be `Iterable` for backwards compatibility

Reply:
Let me rephrase: The step_callback will only ever get called with a list. Is the type signature meant to specify what this function *will get* or *what it can take*?

Reply:
Or we can make it a tuple...

Reply:
ah, so my first statement was a mistake (my last one was accurate but gets inverted here due to functions being contravariant in their argument type), apologies for the confusion!

Here we should use `List`, as you have done, as that gives the most flexibility to the user

Reply:
it does constrain _us_ to pass it a `List`, but I don't envisage that being a problem

Reply:
Ah. Maybe we should define it to be a tuple (which signals some sort of immutability) - actually changing the list that gets passed in would probably lead to some very weird effects??

Reply:
yeah i guess tuple would be safer